include/onnxruntime/core/session/onnxruntime_cxx_api.h

```diff
@@ -210,6 +210,70 @@ struct BFloat16_t {
 
 static_assert(sizeof(BFloat16_t) == sizeof(uint16_t), "Sizes must match");
 
+/** \brief float8e4m3fn (Float8 Floating Point) data type
+ * \details It is necessary for type dispatching to make use of C++ API
+ * The type is implicitly convertible to/from uint8_t.
+ */
+struct Float8E4M3FN_t {
+  uint8_t value;
+  constexpr Float8E4M3FN_t() noexcept : value(0) {}
+  constexpr Float8E4M3FN_t(uint8_t v) noexcept : value(v) {}
+  constexpr operator uint8_t() const noexcept { return value; }
+  // nan values are treated like any other value for operator ==, !=
+  constexpr bool operator==(const Float8E4M3FN_t& rhs) const noexcept { return value == rhs.value; };
+  constexpr bool operator!=(const Float8E4M3FN_t& rhs) const noexcept { return value != rhs.value; };
```

Comment:
I think we want to add operator< for container usage

Reply:
Based on the uint8 used to store the value or the real value is represents?

Reply:
If I understood your question correctly, the comparison should sort it according to the real value semantics, unless this format is transparent to sorting (not familiar with it).

Reply:
It is following ieee standard. The main question is nan. Do you mean < with no exception or < handling nan separatly (nan == nan is false).

Reply:
I was thinking about handling nan == nan is false, since this is what the standard says, but I don't know if these types are standardized.

Reply:
They are.

Reply:
I suggest keeping it that way for the time being. bfloat16 is doing the same, using uint16_t for comparison. I suggest keeping the same for float 8. It seems odd to have bdloat16 equivalent to an int for comparison and float 8 using float standard.

Reply:
You can keep it. Float16 and BFloat16 are being amended to respect NaN in the upcoming PR.